doc/source/whatsnew/v0.23.0.txt

```diff
@@ -1160,6 +1160,7 @@ Reshaping
 - Bug in :meth:`DataFrame.astype` where column metadata is lost when converting to categorical or a dictionary of dtypes (:issue:`19920`)
 - Bug in :func:`cut` and :func:`qcut` where timezone information was dropped (:issue:`19872`)
 - Bug in :class:`Series` constructor with a ``dtype=str``, previously raised in some cases (:issue:`19853`)
+- Stop :func:`concat` and ``Dataframe.append`` from sorting columns by default. Use ``sort=True`` to retain old behavior (:issue:`4588`)
```

Comment:
This should be ``sort=True`` by default to preserve backwards compatibility, right?

Or rather, I think the eventual goal is to have ``sort=False`` be the default, so for now it should be

- ``sort=None`` is the default
- If the default is passed, use ``sort=True`` and warn that the default is changing in the futrue
- If True or False, no warning.

Reply:
actually this needs a sub-section. this is a rather large change (even if its None by default). highliting it is best. pls show an example of previous and new

Reply:
@TomAugspurger
If I do what @jorisvandenbossche suggests then, `sort=True` will not be backwards compatible because it will sort the axes in question regardless of whether the columns are mismatched.
I could have `sort=None` be the default, give a warning and revert to old behavior. In future versions this behavior of only sorting the axes sometimes would not be available because it doesn't make sense and concat could default to `sort=False`

Reply:
@brycepg I think we can do both (it might complicate the code a bit, but not too much I think, as in `_get_combined_index` those cases are already handled separately). As @TomAugspurger suggests, the default can be None for now, so we can raise a warning in the appropriate cases:

- when all axes are equal (current case when there is already no sorting): no warning should be raised, as the future default of `sort=False` will not change anything, but add the ability to also sort the index with `sort=True`
- when not all axes are equal (current case that the unwanted sorting happens): issue a warning that in the future this will no longer sort